nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}:
+</pre>
+
+
+====signMessage====
+Signs a provided messaged with an account selected by user. A randomized salt prefix is added to the input string before it is signed, and it is encased in a non-executable transaction before signed. This ensures allow compatibility with Ledger devices.
+
+
+'''Method Interface'''
+
+<pre>
+function signMessage({message: string}): Promise<SignedMessage>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface SignMessageArgs {
+  message: string; // Arbitrary message to sign
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface SignedMessage {
+  publicKey: string; // Public key of account that signed message
+  message: string; // Original message signed
+  salt: string; // Salt added to original message as prefix, before signing
+  data; string; // Signed message
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string;
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+signMessage({
+  message: 'Hello World!',
+})
+.then((signedMessage: SignedMessage) => {
+  const {
+    publicKey,
+    message,
+    salt,
+    data,
+  } = signedMessage;
+
+  console.log('Public key used to sign:', publicKey);
+  console.log('Original message:', message);
+  console.log('Salt added to message:', salt);
+  console.log('Signed data:', data);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case UNKNOWN_ERROR:
+      console.log(description);
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Present the message to the user for signing
+* Create a randomized salt value to be prefixed to the message before signing
+* Create the hexstring to be signed. '010001f0' + {length of salt+message in hex} + salt + message + '0000'
+* Sign the hexstring value once the user confirms
+* Return to dapp
+
+For more details, please see:
+[https://docs.switcheo.network/#signing-messages-for-neo]
+
+Example
+<pre>
+{
+  publicKey: '0241392007396d6ef96159f047967c5a61f1af0871ecf9dc82afbedb68afbb949a',
+  data: '0147fb89d0999e9d8a90edacfa26152fe695ec8b3770dcad522048297ab903822e12472364e254ff2e088fc3ebb641cc24722c563ff679bb1d1623d08bd5863d0d',
+  salt: '058b9e03e7154e4db1e489c99256b7fa',
+  message: 'Hello World!',
+}
+</pre>
+
+====deploy====
+Will deploy a compiled smart contract to the blockchain with the provided input parameters. The GAS cost for deploying the contract will be calculated by the provider, and displayed to the user upon tx acceptance or rejection.
+
+
+'''Method Interface'''
+
+<pre>
+function deploy({
+  name: string,
+  version: string,
+  author: string,
+  email: string,
+  description: string,
+  needsStorage?: boolean,
```

Comment:
Perhaps these features can stay in a struct named `features` here... like `features : { storage : true, dynamicInvoke : true, payable : true }`, which is easier to extend in the future (or personalize by protocol, Neo2, Ont, Neo3, ...).

Reply:
How is updating attributes in a features struct any different than updating the same on the main object. In either case, such attributes will have to be added or removed, and checked for by name.

Reply:
Just for better organization perhaps.... contract.info.name , ... contract.features.hasStorage ...